ex/loqui/lib/loqui/client.ex

```diff
@@ -0,0 +1,358 @@
+  end
+
+  # Private
+
+  defp handle_packets([], state),
+    do: state
+  defp handle_packets([packet | rest], state) do
+    state = handle_packet(packet, state)
+    handle_packets(rest, state)
+  end
+
+  defp handle_packet({:ping, _flags, seq}, %State{sock: sock}=state) do
+    :gen_tcp.send(sock, Frames.pong(0, seq))
+    state
+  end
+
+  defp handle_packet({:pong, _flags, seq}, %State{last_ping: {seq, _ts}}=state) do
+    %State{state | last_ping: nil}
+  end
+
+  defp handle_packet({:pong, _flags, seq}, %State{last_ping: nil, waiters: waiters}=state) do
+    new_waiters =
+      case Map.pop(waiters, seq) do
+        {nil, waiters} ->
+          Logger.error("Got an unknown pong for sequence #{inspect seq}")
+          waiters
+        {reply_to, waiters} ->
+          Connection.reply(reply_to, :pong)
+          waiters
+      end
+
+    %State{state | waiters: new_waiters}
+  end
+
+  defp handle_packet({:hello_ack, _flags, ping_interval, data}, %State{}=state) do
+    [encoding, compression] = String.split(data, "|")
+    codec = Map.get(state.registered_codecs, encoding)
+    compressor = Map.get(state.registered_compressors, compression)
+
+    %State{state | codec: codec, compressor: compressor, ping_interval: ping_interval}
+      |> schedule_ping()
+  end
+
+  defp handle_packet({:response, _flags, sequence, payload}, %State{waiters: waiters, codec: codec}=state) do
+    {waiter, new_waiters} = Map.pop(waiters, sequence)
+    decoded_data = codec.decode(payload)
+    Connection.reply(waiter, decoded_data)
+
+    %State{state | waiters: new_waiters}
+  end
+
+  defp handle_packet({:go_away, _flags, close_code, payload_data}, %State{}=state) do
+    Process.send_after(self(), {:close_go_away, close_code, payload_data}, @go_away_timeout)
+    %State{state | sequence: :go_away}
+  end
+
+  defp handle_packet(packet, state) do
+    Logger.error("Received unknown packet, opcode #{inspect elem(packet, 0)} #{inspect packet}")
+    state
+  end
+
+  defp schedule_ping(%State{ping_interval: ping_interval}=state) do
+    Process.send_after(self(), :send_ping, ping_interval)
+    state
+  end
+
+  defp make_active_once(%State{sock: sock}=state) do
+    :ok = :inet.setopts(sock, [active: :once])
+
+    state
+  end
+
+  defp do_upgrade(%State{sock: sock, tcp_opts: tcp_opts}=state) do
+    recv_timeout = Keyword.get(tcp_opts, :recv_timeout, @default_timeout)
+    with :ok         <- :gen_tcp.send(sock, upgrade_request(state)),
+         {:ok, data} <- :gen_tcp.recv(sock, 0, recv_timeout),
+         :upgraded   <- parse_upgrade_response(data) do
+
+      {:ok, state}
+    else
+      {:error, reason} ->
+        Logger.error("Upgrade failed #{inspect reason}")
+        {:error, {:upgrade_failed, reason}}
+    end
+  end
+
+  defp to_host(host) when is_bitstring(host),
+    do: String.to_charlist(host)
+
+  defp to_host(host) when is_list(host),
+    do: host
+
+  defp next_sequence(%State{sequence: seq}=state) when seq > @max_sequence do
```

Comment:
Doesn't that put the sequence at `max + 1` when it fails through to the function clause that adds 1 to it?